numba/cuda/simulator/kernel.py

```diff
@@ -160,10 +162,17 @@ def run(self):
             self.exception = (type(e), type(e)(msg), tb)
 
     def syncthreads(self):
+
+        if self.abort:
```

Comment:
I'm not sure why this is needed.  The new test seem to pass regardless of this abort check.

Reply:
Good catch, the failure this flag is intended to address was not addressed in the test case. 

The failure occurs when multiple `syncthreads` calls are made after that statement that raises an exception. In that case the [`syncthreads_event.set()` call in `kernel.py`](https://github.com/numba/numba/pull/3067/files/08728f514a063ad6389906d8ec201fe97b1d5cbf#diff-96f40834dc5291d7b2d5c00ac467ff31R252) will release the blocked kernel threads, which will then re-block on the subsequent syncthreads call. The test case has been expanded in 4d0b9fc2428c55b3170b775865cd8e9e43e0c579 to address this issue.

You'll notice that the test will likely *still pass* after removing these abort statements in the updated test case. This is due to a race condition during simulator shutdown, which *may* occur if the abort flag is not in place:

1. An exception is raised in a kernel thread, triggering the [block-level exception handler](https://github.com/numba/numba/pull/3067/files/08728f514a063ad6389906d8ec201fe97b1d5cbf#diff-96f40834dc5291d7b2d5c00ac467ff31R248). The handler (event A) unblocks all other kernel threads and (event B) reraises the exception, [unshimming the `cuda.syncthreads` function added via `swapped_cuda_module`](https://github.com/uw-ipd/numba/blob/08728f514a063ad6389906d8ec201fe97b1d5cbf/numba/cuda/simulator/kernel.py#L91).

2. Previously blocked kernel threads awake after event A and begin processing until (event C) arriving at the next `syncthreads` call.

If B occurs *before* C (the most likely case), the `syncthreads` function has been unshimmed, and the call to `cuda.syncthreads` fails with an attribute error. If C occurs *before* B (an unlikely case), the `syncthreads` function will not yet be unshimmed, and the call to `cuda.syncthreads` will reblock the processing thread. 

You can trigger this condition (C before B), causing a test failure *if the abort flag is not checked or set* , by inserting a call to `sleep` before re-raising the exception:

```diff
diff --git a/numba/cuda/simulator/kernel.py b/numba/cuda/simulator/kernel.py
index ed2933c..e662cb7 100644
--- a/numba/cuda/simulator/kernel.py
+++ b/numba/cuda/simulator/kernel.py
@@ -246,10 +246,12 @@ class BlockManager(object):

                     # Abort all other simulator threads on exception,
                     # do *not* join immediately to facilitate debugging.
+                    import time
                     for t_other in threads:
-                        t_other.abort = True
                         t_other.syncthreads_blocked = False
                         t_other.syncthreads_event.set()
+                    time.sleep(1)

                     reraise(*(t.exception))
             if livethreads == blockedthreads:
```